analytic_engine/src/setup.rs

```diff
@@ -71,17 +71,30 @@ const WAL_DIR_NAME: &str = "wal";
 const MANIFEST_DIR_NAME: &str = "manifest";
 const STORE_DIR_NAME: &str = "store";
 
-type InstanceRefOnTableKv<T> = InstanceRef<WalNamespaceImpl<T>, ManifestImpl<WalNamespaceImpl<T>>>;
-
 /// Analytic engine builder.
 #[async_trait]
-pub trait EngineBuilder: Default {
+pub trait EngineBuilder: Send + Sync + Default {
     /// Build the analytic engine from `config` and `engine_runtimes`.
     async fn build(
         &self,
         config: Config,
         engine_runtimes: Arc<EngineRuntimes>,
-    ) -> Result<TableEngineRef>;
+    ) -> Result<TableEngineRef> {
+        assert!(!config.obkv_wal.enable);
+
+        let store = open_storage(config.storage.clone()).await?;
+        let (wal, manifest) = self
+            .open_wal_and_manifest(config.clone(), engine_runtimes.clone())
+            .await?;
+        let instance = open_instance(config.clone(), engine_runtimes, wal, manifest, store).await?;
+        Ok(Arc::new(TableEngineImpl::new(instance)))
+    }
```

Comment:
`This default implementation` can be used for all `EngineBuilder`.
I found the three `build` method has the same logic:

1. open staorage.
2. open wal and manifest.
3. open instance.
4. build TableEngineRef.

The difference between these `EngineBuilder` is how to open `wal` and `manifest`, so I `extract` a method named `open_wal_and_manifest`, after that the default `build` can be used for all  `EngineBuilder`, and different `EngineBuilder` only need implement their own `open_wal_and_manifest`.

```rust
 pub trait EngineBuilder: Send + Sync + Default {
    async fn build(&self, config: Config, engine_runtimes: Arc<EngineRuntimes>) -> Result<TableEngineRef> {
        let store = open_storage(config.storage.clone()).await?;
        let (wal, manifest) = self.open_wal_and_manifest(..);
        let instance = open_instance(..);
        Ok(Arc::new(TableEngineImpl::new(instance)))
    }

    async fn open_wal_and_manifest(&self, config: Config,engine_runtimes: Arc<EngineRuntimes>,
    ) -> Result<(WalManagerRef, ManifestImpl)>;
}
```

Reply:
Got it. This is a good approach 👍 

I misread the assertion in L83, so it's just a redundant part 😃

Reply:
And I think maybe we need to remove this line? 
https://github.com/CeresDB/ceresdb/pull/163/files#diff-5cabf23fa80a7e240c564b81b96c954a15f319d71b88aaaf33697c7c31e91514R83
```rust
        assert!(!config.obkv_wal.enable);
```

Reply:
Seems we can move assert to `open_wal_and_manifest` now, because the original code will check this.

`RocksEngineBuilder`:
https://github.com/CeresDB/ceresdb/blob/cb96b5ff100e5966bbaf5a9045abb279144c6394/analytic_engine/src/setup.rs#L98

`ReplicatedEngineBuilder`:
https://github.com/CeresDB/ceresdb/blob/cb96b5ff100e5966bbaf5a9045abb279144c6394/analytic_engine/src/setup.rs#L117

`MemWalEngineBuilder` do not `assert!`.

Reply:
This sounds good!